rfcs/proposed/flow_graph_serializers/README.md

```diff
@@ -0,0 +1,596 @@
+
+1.  placing an explicit lock around the use of the database handle,
+    resulting in inefficiencies in the execution of the graph,
+2.  creating explicit edges between `track_maker` and `cluster_maker`
+    even though there is no obvious data dependency between them,
+3.  creating a token-based system that can limit access to a shared
+    resource.
+
+This RFC proposes an interface that pursues option 3, which we describe
+in the “Implementation experience” section, below. Our proposal,
+however, does not mandate any implementation but suggests an API similar
+to:
+
+``` cpp
+using namespace tbb;
+flow::graph g;
+
+resource_limiter_node<DB> db_resource{g, {db_handle()}};
+
+rl_function_node<Hits, Tracks> track_maker{
+  g,
+  std::tie(db_resource),
+  [](Hits const&, DB const* db) -> Tracks { auto a = db->access(/* okay */); ... }
+};
+rl_function_node<Signals, Clusters> cluster_maker{
+  g,
+  std::tie(db_resource),
+  [](Signals const&, DB const* db) -> Clusters { auto a = db->access(/* okay */); ... }
+};
+```
+
+where `db_resource` ensures limited access to a resource that both
+`track_maker` and `cluster_maker` require. The implementation of the
+`DB::access()` function did not change, but by connecting the
+resource-limited function nodes to the `db_resource`, the bodies of the
+nodes will not be invoked concurrently. Note that if the only reason
+that the bodies of `track_maker` and `cluster_maker` were thread-unsafe
+was their access to the limited resource indicated by `db_resource` it
+is no longer necessary to declare that the nodes have concurrency
+`flow::serial`. It may be possible to have the node `track_maker` active
+at the same time, if the nature of `db_resource` were to allow two
+handles to be available, and as long as each activation was given a
+token to a different handle.
+
+# Proposal
+
+Our proposal is an addition to what already exists and does not break
+API backwards compatibility. The proposal consists of: 1. Introducing
+the equivalent of a `resource_limiter_node` class template that, when
+connected with another node, ensures limited access to the resource it
+represents. 2. Adding a `rl_function_node` class template that allow the
+specification of limited resource nodes instead of (or in addition to) a
+`concurrency` value.
+
+> \[!NOTE\] Although we pattern our proposal on the
+> `flow::function_node` class template in this proposal, the concepts
+> discussed here apply to nearly any flow-graph node that accepts a
+> user-provided node body.
+
+With the proposal here, in addition to what is written above we imagine
+the following could be done:
+
+``` cpp
+using namespace tbb;
+flow::graph g;
+
+resource_limiter_node<GPU> gpus{g, {GPU{ /*gpu 1*/ }, GPU{ /*gpu 2*/}}; // Permit access to two GPUs
+resource_limiter_node<ROOT> root{g, {ROOT{...}}};  // Only 1 ROOT handle available
+
+rl_function_node fn{g,
+                    std::tie(gpus, root), // Edges implicitly created to the required resource limiters
```

Comment:
It seems a bit strange to me to define `gpus` and `root` as nodes. As a user, I would expect a node in the graph as something that can send and/or receive some kind of value and may be connected with other nodes that can send/receive the same value type. 
It is a bit unclear what kind of values can be sent by `gpus` and `root`. They seem more like special instances that manages the resource and the `function_node`s just using the managed resources.
I propose not to look on things like `gpus` and `root` as on the graph nodes and choose a better public name for that. 

In theory, they can be implemented as a graph nodes sending/receiving the token that is used to manage the resource, but I don't think such details should be expressed in the API names.

Reply:
We agree with your point, @kboyarinov.  We will rename `resource_limiter_node` to just `resource_limiter`, even though our implementation expressed it as a Flow Graph node.

Reply:
It turns we also had a footnote in our analysis that anticipated such a change.  That footnote is now unnecessary.